File: output/rust/tests/02_interface.rs
use std::collections::HashMap;

struct WithInterfaceFrameEvent {
    message: String,
    parameters: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl Clone for WithInterfaceFrameEvent {
    fn clone(&self) -> Self {
        Self {
            message: self.message.clone(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

impl WithInterfaceFrameEvent {
    fn new(message: &str) -> Self {
        Self {
            message: message.to_string(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

struct WithInterfaceFrameContext {
    event: WithInterfaceFrameEvent,
    _return: Option<Box<dyn std::any::Any>>,
    _data: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl WithInterfaceFrameContext {
    fn new(event: WithInterfaceFrameEvent, default_return: Option<Box<dyn std::any::Any>>) -> Self {
        Self {
            event,
            _return: default_return,
            _data: std::collections::HashMap::new(),
        }
    }
}

#[derive(Clone)]
struct WithInterfaceCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<WithInterfaceFrameEvent>,
}

impl WithInterfaceCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone)]
enum WithInterfaceStateContext {
    Ready,
    Empty,
}

impl Default for WithInterfaceStateContext {
    fn default() -> Self {
        WithInterfaceStateContext::Ready
    }
}

pub struct WithInterface {
    _state_stack: Vec<(String, WithInterfaceStateContext)>,
    __compartment: WithInterfaceCompartment,
    __next_compartment: Option<WithInterfaceCompartment>,
    _context_stack: Vec<WithInterfaceFrameContext>,
    call_count: i32,
}

impl WithInterface {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _context_stack: vec![],
            call_count: 0,
            __compartment: WithInterfaceCompartment::new("Ready"),
            __next_compartment: None,
        };
let __frame_event = WithInterfaceFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: WithInterfaceFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = WithInterfaceFrameEvent::new("<$");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = WithInterfaceFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = WithInterfaceFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &WithInterfaceFrameEvent) {
match self.__compartment.state.as_str() {
    "Ready" => self._state_Ready(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: WithInterfaceCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Ready" => WithInterfaceStateContext::Ready,
    _ => WithInterfaceStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = WithInterfaceFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    WithInterfaceStateContext::Ready => {}
    WithInterfaceStateContext::Empty => {}
}
    }

    pub fn greet(&mut self, name: String) -> String {
let mut __e = WithInterfaceFrameEvent::new("greet");
__e.parameters.insert("name".to_string(), Box::new(name.clone()) as Box<dyn std::any::Any>);
let __ctx = WithInterfaceFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Ready" => { self._s_Ready_greet(&__e, name); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = WithInterfaceFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = WithInterfaceFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = WithInterfaceFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<String>().unwrap()
} else {
    Default::default()
}
    }

    pub fn get_count(&mut self) -> i32 {
let mut __e = WithInterfaceFrameEvent::new("get_count");
let __ctx = WithInterfaceFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Ready" => { self._s_Ready_get_count(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = WithInterfaceFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = WithInterfaceFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = WithInterfaceFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<i32>().unwrap()
} else {
    Default::default()
}
    }

    fn _state_Ready(&mut self, __e: &WithInterfaceFrameEvent) {
match __e.message.as_str() {
    "get_count" => { self._s_Ready_get_count(__e); }
    _ => {}
}
    }

    fn _s_Ready_greet(&mut self, __e: &WithInterfaceFrameEvent, name: String) {
self.call_count += 1;
if let Some(ctx) = self._context_stack.last_mut() { ctx._return = Some(Box::new(format!("Hello, {}!", name))); }
return;
    }

    fn _s_Ready_get_count(&mut self, __e: &WithInterfaceFrameEvent) {
if let Some(ctx) = self._context_stack.last_mut() { ctx._return = Some(Box::new(self.call_count)); }
return;
    }
}


fn main() {
    println!("=== Test 02: Interface Methods ===");
    let mut s = WithInterface::new();

    // Test interface method with parameter and return
    let result = s.greet("World".to_string());
    assert_eq!(result, "Hello, World!", "Expected 'Hello, World!', got '{}'", result);
    println!("greet('World') = {}", result);

    // Test domain variable access through interface
    let count = s.get_count();
    assert_eq!(count, 1, "Expected count=1, got {}", count);
    println!("get_count() = {}", count);

    // Call again to verify state
    s.greet("Frame".to_string());
    let count2 = s.get_count();
    assert_eq!(count2, 2, "Expected count=2, got {}", count2);
    println!("After second call: get_count() = {}", count2);

    println!("PASS: Interface methods work correctly");
}

File: output/go/tests/02_interface.go

package main

import "fmt"
import "os"

type WithInterfaceFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type WithInterfaceFrameContext struct {
    _event  WithInterfaceFrameEvent
    _return any
    _data   map[string]any
}

type WithInterfaceCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *WithInterfaceFrameEvent
    parentCompartment *WithInterfaceCompartment
}

func newWithInterfaceCompartment(state string) *WithInterfaceCompartment {
    return &WithInterfaceCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *WithInterfaceCompartment) copy() *WithInterfaceCompartment {
    nc := &WithInterfaceCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type WithInterface struct {
    _state_stack []*WithInterfaceCompartment
    __compartment *WithInterfaceCompartment
    __next_compartment *WithInterfaceCompartment
    _context_stack []WithInterfaceFrameContext
    call_count int
}

func NewWithInterface() *WithInterface {
    s := &WithInterface{}
    s._state_stack = make([]*WithInterfaceCompartment, 0)
    s._context_stack = make([]WithInterfaceFrameContext, 0)
    s.__compartment = newWithInterfaceCompartment("Ready")
    s.__next_compartment = nil
    __frame_event := WithInterfaceFrameEvent{_message: "$>", _parameters: nil}
    __ctx := WithInterfaceFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *WithInterface) __kernel(__e *WithInterfaceFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &WithInterfaceFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &WithInterfaceFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &WithInterfaceFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *WithInterface) __router(__e *WithInterfaceFrameEvent) {
    switch s.__compartment.state {
    case "Ready":
        s._state_Ready(__e)
    }
}

func (s *WithInterface) __transition(next *WithInterfaceCompartment) {
    s.__next_compartment = next
}

func (s *WithInterface) Greet(name string) string {
    __params := map[string]any{
        "name": name,
    }
    __e := WithInterfaceFrameEvent{_message: "Greet", _parameters: __params}
    __ctx := WithInterfaceFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *WithInterface) GetCount() int {
    __e := WithInterfaceFrameEvent{_message: "GetCount"}
    __ctx := WithInterfaceFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *WithInterface) _state_Ready(__e *WithInterfaceFrameEvent) {
    if __e._message == "GetCount" {
        s._context_stack[len(s._context_stack)-1]._return = s.call_count
        return
    } else if __e._message == "Greet" {
        name := __e._parameters["name"].(string)
        _ = name
        s.call_count += 1
        s._context_stack[len(s._context_stack)-1]._return = "Hello, " + name + "!"
        return
    }
}

func main() {
	fmt.Println("=== Test 02: Interface Methods ===")
	sm := NewWithInterface()

	// Test interface method with parameter and return
	result := sm.Greet("World")
	if result != "Hello, World!" {
		fmt.Printf("FAIL: Expected 'Hello, World!', got '%s'\n", result)
		os.Exit(1)
	}
	fmt.Printf("Greet('World') = %s\n", result)

	// Test domain variable access through interface
	count := sm.GetCount()
	if count != 1 {
		fmt.Printf("FAIL: Expected count=1, got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("GetCount() = %d\n", count)

	// Call again to verify state
	sm.Greet("Frame")
	count2 := sm.GetCount()
	if count2 != 2 {
		fmt.Printf("FAIL: Expected count=2, got %d\n", count2)
		os.Exit(1)
	}
	fmt.Printf("After second call: GetCount() = %d\n", count2)

	fmt.Println("PASS: Interface methods work correctly")
}
